server/tserver/src/main/java/org/apache/accumulo/tserver/ScanServer.java

```diff
@@ -0,0 +1,896 @@
+    }
+
+    ServiceLock lock = announceExistence();
+
+    try {
+      MetricsUtil.initializeMetrics(getContext().getConfiguration(), this.applicationName,
+          clientAddress);
+    } catch (Exception e1) {
+      LOG.error("Error initializing metrics, metrics will not be emitted.", e1);
+    }
+    scanMetrics = new TabletServerScanMetrics();
+    MetricsUtil.initializeProducers(scanMetrics);
+
+    // We need to set the compaction manager so that we don't get an NPE in CompactableImpl.close
+    ceMetrics = new CompactionExecutorsMetrics();
+    this.compactionManager = new ScanServerCompactionManager(getContext(), ceMetrics);
+
+    // SessionManager times out sessions when sessions have been idle longer than
+    // TSERV_SESSION_MAXIDLE. Create a background thread that looks for sessions that
+    // have timed out and end the scan.
+    long maxIdle =
+        this.getContext().getConfiguration().getTimeInMillis(Property.TSERV_SESSION_MAXIDLE);
+    LOG.debug("Looking for timed out scan sessions every {}ms", Math.max(maxIdle / 2, 1000));
+    this.getContext().getScheduledExecutor().scheduleWithFixedDelay(() -> cleanupTimedOutSession(),
+        Math.max(maxIdle / 2, 1000), Math.max(maxIdle / 2, 1000), TimeUnit.MILLISECONDS);
+
+    try {
+      try {
+        ScanServerDiscovery.unreserve(this.getContext().getZooKeeperRoot(),
+            getContext().getZooReaderWriter(), getClientAddressString());
+      } catch (Exception e2) {
+        throw new RuntimeException("Error setting initial unreserved state in ZooKeeper", e2);
+      }
+
+      while (!serverStopRequested) {
+        UtilWaitThread.sleep(1000);
+      }
+    } finally {
+      LOG.info("Stopping Thrift Servers");
+      TServerUtils.stopTServer(address.server);
+
+      try {
+        LOG.debug("Closing filesystems");
+        VolumeManager mgr = getContext().getVolumeManager();
+        if (null != mgr) {
+          mgr.close();
+        }
+      } catch (IOException e) {
+        LOG.warn("Failed to close filesystem : {}", e.getMessage(), e);
+      }
+
+      gcLogger.logGCInfo(getConfiguration());
+      LOG.info("stop requested. exiting ... ");
+      try {
+        if (null != lock) {
+          lock.unlock();
+        }
+      } catch (Exception e) {
+        LOG.warn("Failed to release scan server lock", e);
+      }
+
+    }
+  }
+
+  private synchronized void checkInUse() {
+    synchronized (currentScan) {
+      if (currentScan.extent != null || currentScan.tablet != null) {
+        throw new RuntimeException("Scan server in use for another query");
+      }
+    }
+  }
+
+  protected synchronized boolean loadTablet(TKeyExtent textent)
+      throws IllegalArgumentException, IOException, AccumuloException {
+    synchronized (currentScan) {
+      KeyExtent extent = KeyExtent.fromThrift(textent);
+      TabletMetadata tabletMetadata = getContext().getAmple().readTablet(extent);
```

Comment:
@dlmarion when we talked earlier today we discussed that the Tablet or TabletMetadata could be cached for a limited time.  Something for one of us to look into  eventually.

Reply:
This has been added, see `tabletMetadataCache` in ScanServer.